flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/remote/RemoteStorageScanner.java

```diff
@@ -0,0 +1,241 @@
+        this.notifier = retriever;
+    }
+
+    // ------------------------------------------------------------------------
+    //  Internal Methods
+    // ------------------------------------------------------------------------
+
+    /**
+     * Scan the max segment id of segment files for the specific partition and subpartition. The max
+     * segment id can be obtained from a file named by max segment id.
+     *
+     * @param partitionId the partition id.
+     * @param subpartitionId the subpartition id.
+     */
+    private void scanMaxSegmentId(
+            TieredStoragePartitionId partitionId, TieredStorageSubpartitionId subpartitionId) {
+        Path segmentFinishDir =
+                getSegmentFinishDirPath(
+                        baseRemoteStoragePath, partitionId, subpartitionId.getSubpartitionId());
+        FileStatus[] fileStatuses;
+        try {
+            if (!remoteFileSystem.exists(segmentFinishDir)) {
+                return;
+            }
+            fileStatuses = remoteFileSystem.listStatus(segmentFinishDir);
+        } catch (IOException e) {
+            throw new RuntimeException(
+                    "Failed to list the segment finish file. " + segmentFinishDir, e);
+        }
+        if (fileStatuses.length == 0) {
+            return;
+        }
+        checkState(fileStatuses.length == 1);
+        scannedMaxSegmentIds.put(
+                Tuple2.of(partitionId, subpartitionId),
+                Integer.parseInt(fileStatuses[0].getPath().getName()));
+    }
+
+    /**
+     * The strategy is used to decide the scan interval of {@link RemoteStorageScanner}. The
+     * interval will be updated exponentially and restricted by max value.
+     */
+    private static class ScanStrategy {
+
+        private final int initialScanInterval;
+
+        private final int maxScanInterval;
+
+        private ScanStrategy(int initialScanInterval, int maxScanInterval) {
+            checkArgument(
+                    initialScanInterval > 0,
+                    "initialScanInterval must be positive, was %s",
+                    initialScanInterval);
+            checkArgument(
+                    maxScanInterval > 0,
+                    "maxScanInterval must be positive, was %s",
+                    maxScanInterval);
+            checkArgument(
+                    initialScanInterval <= maxScanInterval,
+                    "initialScanInterval must be lower than or equal to maxScanInterval",
+                    maxScanInterval);
+            this.initialScanInterval = initialScanInterval;
+            this.maxScanInterval = maxScanInterval;
+        }
+
+        private long getInterval(int attempt) {
+            checkArgument(attempt >= 0, "attempt must not be negative (%s)", attempt);
+            final long interval = initialScanInterval * Math.round(Math.pow(2, attempt));
+            return interval >= 0 && interval < maxScanInterval ? interval : maxScanInterval;
+        }
```

Comment:
Relying on `interval >= 0` to check for overflows can be problematic, because the `attempt` is keep growing and it's possible that an overflowed interval may still fall into the range `[0, maxScanInterval)`. Alternatively, you may replace `attempt` with a `lastInterval`, which stops growing after reaching the max.